crates/sui-graphql-rpc/src/extensions/timeout.rs

```diff
@@ -2,35 +2,69 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use async_graphql::{
-    extensions::{Extension, ExtensionContext, ExtensionFactory, NextRequest},
-    Response, ServerError,
+    extensions::{Extension, ExtensionContext, ExtensionFactory, NextExecute, NextParseQuery},
+    parser::types::ExecutableDocument,
+    Response, ServerError, ServerResult,
 };
-use std::sync::Arc;
+use async_graphql_value::Variables;
 use std::time::Duration;
-use tokio::time::timeout;
+use std::{net::SocketAddr, sync::Arc};
+use tokio::{sync::Mutex, time::timeout};
+use tracing::error;
+use uuid::Uuid;
 
-use crate::config::ServiceConfig;
+use crate::{config::ServiceConfig, error::code};
 
-#[derive(Clone, Debug, Default)]
-pub(crate) struct Timeout;
+#[derive(Debug, Default)]
+pub(crate) struct Timeout {
+    pub query: Mutex<Option<String>>,
+}
 
 impl ExtensionFactory for Timeout {
     fn create(&self) -> Arc<dyn Extension> {
-        Arc::new(Timeout)
+        Arc::new(Timeout {
+            query: Mutex::new(None),
+        })
     }
 }
 
 #[async_trait::async_trait]
 impl Extension for Timeout {
-    async fn request(&self, ctx: &ExtensionContext<'_>, next: NextRequest<'_>) -> Response {
+    async fn parse_query(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        query: &str,
+        variables: &Variables,
+        next: NextParseQuery<'_>,
+    ) -> ServerResult<ExecutableDocument> {
+        let document = next.run(ctx, query, variables).await?;
+        *self.query.lock().await = Some(ctx.stringify_execute_doc(&document, variables));
+        Ok(document)
+    }
+
+    async fn execute(
+        &self,
+        ctx: &ExtensionContext<'_>,
+        operation_name: Option<&str>,
+        next: NextExecute<'_>,
+    ) -> Response {
         let cfg = ctx
             .data::<ServiceConfig>()
             .expect("No service config provided in schema data");
+        let query = self.query.lock().await.take().unwrap_or("".to_string());
         let request_timeout = Duration::from_millis(cfg.limits.request_timeout_ms);
-
-        timeout(request_timeout, next.run(ctx))
+        timeout(request_timeout, next.run(ctx, operation_name))
             .await
             .unwrap_or_else(|_| {
+                let query_id: &Uuid = ctx.data_unchecked();
+                let session_id: &SocketAddr = ctx.data_unchecked();
+                let error_code = code::REQUEST_TIMEOUT.to_string();
```

Comment:
Yes, I think we can use our own `graphql_error_response` that wraps `Response::from_errors` and includes the `error_code`.

Reply:
Btw, I am not sure I follow why it would be a breaking change? What we return now is the following. We add error_code = "UNKNOWN" to all errors that do not have an `error_code` set, so instead this would become REQUEST_TIMEOUT. I'm probably missing something..
```
2024-02-29T17:48:37.265024Z  INFO sui_graphql_rpc::extensions::logger: [Response] message=Request timed out. Limit: 0.5s query_id=8cddf7c6-370a-4de5-92fe-bee48dc23bf6 session_id=127.0.0.1:61093 error_code="UNKNOWN"
```

Reply:
Changing the error code in the log message is not a breaking change, but changing the error code that we return to users would be.